core/src/queue.rs

```diff
@@ -304,12 +306,10 @@ impl Queue {
             return;
         }
 
-        let mut seen_queue = Vec::new();
         let mut expired_transactions = Vec::new();
 
-        let txs_from_queue = core::iter::from_fn(|| {
-            self.pop_from_queue(&mut seen_queue, state_view, &mut expired_transactions)
-        });
+        let txs_from_queue =
+            core::iter::from_fn(|| self.pop_from_queue(state_view, &mut expired_transactions));
```

Comment:
```suggestion
        let txs_from_queue =
            core::iter::from_fn(|| self.pop_from_queue(state_view, &mut expired_transactions));
```

could `pop` return a `TransactionGuard` instead? so that when the guard is dropped the transaction is also removed from the queue. I suppose `TransactionGuard` would keep a reference to the `Queue` and have an implementation of `Drop` that calls `remove_stale_transaction`